libs/internals/core.write_compiled_resource.php

```diff
@@ -14,7 +43,18 @@
  */
 function smarty_core_write_compiled_resource($params, &$smarty)
 {
-    if(!@is_writable($smarty->compile_dir)) {
+    /**
+     * Detect if running on a Windows server
+     * Added 04/03/2015 by Yola. Necessary to avoid bug in PHP with
+     * is_writable() not working correctly with ACL permissions.
+     */
+    $isWin = strtoupper(substr(php_uname('s'), 0, 3)) === 'WIN';
```

Comment:
Why don't you move this logic into your is_writeable function? Then we just call it, everywhere, and it wraps a call to PHP's builtin, if it thinks it'll work?